pageserver/src/pgdatadir_mapping.rs

```diff
@@ -0,0 +1,1351 @@
+
+        Ok(())
+    }
+
+    pub fn put_slru_segment_creation(
+        &mut self,
+        kind: SlruKind,
+        segno: u32,
+        nblocks: BlockNumber,
+    ) -> Result<()> {
+        // Add it to the directory entry
+        let dir_key = slru_dir_to_key(kind);
+        let buf = self.get(dir_key)?;
+        let mut dir = SlruSegmentDirectory::des(&buf)?;
+
+        if !dir.segments.insert(segno) {
+            bail!("slru segment {:?}/{} already exists", kind, segno);
+        }
+        self.put(
+            dir_key,
+            Value::Image(Bytes::from(SlruSegmentDirectory::ser(&dir)?)),
+        );
+
+        // Put size
+        let size_key = slru_segment_size_to_key(kind, segno);
+        let buf = nblocks.to_le_bytes();
+        self.put(size_key, Value::Image(Bytes::from(buf.to_vec())));
+
+        // even if nblocks > 0, we don't insert any actual blocks here
+
+        Ok(())
+    }
+
+    /// Extend SLRU segment
+    pub fn put_slru_extend(
+        &mut self,
+        kind: SlruKind,
+        segno: u32,
+        nblocks: BlockNumber,
+    ) -> Result<()> {
+        // Put size
+        let size_key = slru_segment_size_to_key(kind, segno);
+        let buf = nblocks.to_le_bytes();
+        self.put(size_key, Value::Image(Bytes::from(buf.to_vec())));
+        Ok(())
+    }
+
+    /// This method is used for marking truncated SLRU files
+    pub fn drop_slru_segment(&mut self, kind: SlruKind, segno: u32) -> Result<()> {
+        // Remove it from the directory entry
+        let dir_key = slru_dir_to_key(kind);
+        let buf = self.get(dir_key)?;
+        let mut dir = SlruSegmentDirectory::des(&buf)?;
+
+        if !dir.segments.remove(&segno) {
+            warn!("slru segment {:?}/{} does not exist", kind, segno);
+        }
+        self.put(
+            dir_key,
+            Value::Image(Bytes::from(SlruSegmentDirectory::ser(&dir)?)),
+        );
+
+        // Delete size entry, as well as all blocks
+        self.delete(slru_segment_key_range(kind, segno));
+
+        Ok(())
+    }
+
+    /// This method is used for marking dropped relations and truncated SLRU files and aborted two phase records
+    pub fn drop_relmap_file(&mut self, _spcnode: Oid, _dbnode: Oid) -> Result<()> {
+        // TODO
+        Ok(())
+    }
+
+    /// This method is used for marking truncated SLRU files
+    pub fn drop_twophase_file(&mut self, xid: TransactionId) -> Result<()> {
+        // Remove it from the directory entry
+        let buf = self.get(TWOPHASEDIR_KEY)?;
+        let mut dir = TwoPhaseDirectory::des(&buf)?;
+
+        if !dir.xids.remove(&xid) {
+            warn!("twophase file for xid {} does not exist", xid);
+        }
+        self.put(
+            TWOPHASEDIR_KEY,
+            Value::Image(Bytes::from(TwoPhaseDirectory::ser(&dir)?)),
+        );
+
+        // Delete it
+        self.delete(twophase_key_range(xid));
+
+        Ok(())
+    }
+
+    pub fn finish(self) -> Result<()> {
```

Comment:
Please, add a comment. 
Maybe make `begin_record` & `finish` names matching? It would be easier to understand that they come together.

Reply:
I renamed the struct to `DatadirModification`, to make it clear that it represents one atomic modification. And renamed the functions to `begin_modification` and just `commit`.

One downside the term "modification" is that you cannot have a local variable called "mod", because that's a keyword in rust. So this doesn't work:

```
let mod = tline.begin_modification(lsn);
mod.put...()
mod.commit();
```
Annoyingly, the term "mutation" has the same problem, "mut" is also a Rust keyword :-). So I renamed any such local variables to "modification" or just "m" if it's short-lived.